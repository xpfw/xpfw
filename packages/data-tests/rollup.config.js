// rollup.config.js
import typescript from "rollup-plugin-typescript2"
import uglify from "rollup-plugin-uglify"
import { minify } from "uglify-es"

export default {
  input: 'src/index.ts',
  output: {
    file: 'dist/index.js',
    format: 'cjs'
  },
  plugins: [
    typescript(),
    // uglify({}, minify)
  ],
  external: [
    "isomorphic-fetch",
    "ajv",
    "lodash",
    "lodash-es",
    "mobx",
    "react",
    "react-dom",
    "preact-render-to-json",
    "moment",
    "mockdate",
    "@xpfw/form",
    "@xpfw/form-tests",
    "@xpfw/form-shared",
    "@xpfw/ui-shared",
    "moment",
    "@feathersjs/feathers",
    "@feathersjs/authentication-client",
    "@feathersjs/socketio-client",
    "@feathersjs/rest-client",
    "resub",
    "resub-persist",
    "socket.io-client",
    "@xpfw/validate",
    "@xpfw/form-shared",
    "@xpfw/ui-tests",
    "@xpfw/ui-feathers",
    "@xpfw/data",
    "@xpfw/data-tests",
    "@xpfw/data-feathers",
    "@xpfw/test-util",
    "react",
    "react-dom",
    "react-test-renderer",
    "isomorphic-fetch"
  ]
};