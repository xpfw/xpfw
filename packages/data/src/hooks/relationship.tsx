import { ExtendedJSONSchema, FormStore, getMapTo, IFieldProps, memo, prependPrefix, useField, useFieldWithValidation } from "@xpfw/form"
import { cloneDeep, get, isArray, isNil, isNumber, isString, remove } from "lodash"
import { action } from "mobx"
import * as React from "react"
import dataOptions from "../options"
import DbStore from "../store/db"
import ListStore from "../store/list"

const lastUsedKey = "lastUsed"

const getListFormFromRelationshipField:
(schema: ExtendedJSONSchema, mapTo?: string) => ExtendedJSONSchema = (schema, mapTo) => {
  const collection = get(schema, "relationship.collection")
  const nameTransform: any = get(schema, "relationship.nameTransform", (schema: any, val: any) => {
    if (val == null || val.length === 0) {
      return null
    }
    return {
      $regex: `(.*?)${val}(.*?)`,
      $options: "isg"
    }
  })
  const nameSearchField: ExtendedJSONSchema = {
    title: get(schema, "relationship.namePath"),
    type: "string"
  }
  return {
    title: "searchFor." + getMapTo(schema, mapTo),
    collection,
    properties: {
      [String(nameSearchField.title)]: nameSearchField,
      [dataOptions.idPath]: {
        title: dataOptions.idPath,
        type: "string"
      }
    },
    relationship: {
      nameTransform
    }
  }
}

const displayModeChanger = (mapTo: string, prefix?: string) => {
  return (newValue: any) => {
    FormStore.setValue(`displayMode.${mapTo}`, newValue, prefix)
  }
}

const addId = (schema: ExtendedJSONSchema, mapTo?: string, prefix?: string) => {
  if (mapTo == null) { mapTo = getMapTo(schema, mapTo) }
  const disp = displayModeChanger(mapTo, prefix)
  return action((newValue: any) => {
    if (schema.type === "array") {
      let currentValue = FormStore.getValue(mapTo, prefix)
      currentValue = !isArray(currentValue) ? [] : currentValue
      currentValue.push(newValue)
      FormStore.setValue(mapTo, currentValue, prefix)
      disp(0)
    } else {
      FormStore.setValue(mapTo, newValue, prefix)
    }
    const lastUsedFieldKey = `${lastUsedKey}.${prependPrefix(mapTo, prefix)}`
    let lastUsed = FormStore.getValue(lastUsedFieldKey)
    if (!Array.isArray(lastUsed)) {
      lastUsed = []
    }
    const indexOfLastUsed = lastUsed.indexOf(newValue)
    if (indexOfLastUsed !== -1) {
      lastUsed.splice(indexOfLastUsed, 1)
    }
    lastUsed.unshift(newValue)
    if (lastUsed.length > ListStore.pageSize) {
      lastUsed.splice(1)
    }
    FormStore.setValue(lastUsedFieldKey, lastUsed)
  })
}

const removeId = (schema: ExtendedJSONSchema, mapTo?: string, prefix?: string) => {
  if (mapTo == null) { mapTo = getMapTo(schema, mapTo) }
  return action((newValue: any) => {
    if (schema.type === "array") {
      let currentValue = FormStore.getValue(mapTo)
      currentValue = !Array.isArray(currentValue) ? [] : cloneDeep(currentValue)
      remove(currentValue, (val) => val === newValue)
      FormStore.setValue(mapTo, currentValue, prefix)
    } else {
      FormStore.setValue(mapTo, undefined, prefix)
    }
  })
}

const searchRelated = (schema: ExtendedJSONSchema, mapTo?: string, prefix?: string) => {
  if (mapTo == null) { mapTo = getMapTo(schema, mapTo) }
  return async (newValue: any) => {
    const form = getListFormFromRelationshipField(schema)
    const value = FormStore.getValue(mapTo, prefix)
    if (get(schema.relationship, "filterOutSelected", false) === true) {
      const f = useField(dataOptions.idPath, prependPrefix(mapTo, prefix))
      f.setValue({$nin: Array.isArray(value) ? value : [value]})
    }
    const nameField = useField(get(schema, "relationship.namePath"), prependPrefix(form.title, prefix))
    const transformer = get(form, "relationship.nameTransform")
    nameField.setValue(transformer(schema, newValue))
    const res: any = await ListStore.getList(form, undefined, prefix, true)
    if (get(schema.relationship, "autoSelect", false) === true) {
      if (!isNil(res) && Array.isArray(res.data) && res.data.length === 1) {
        addId(schema, mapTo, prefix)(get(res.data[0], dataOptions.idPath))
      }
    }
    return res
  }
}

// public componentDidUpdate() {
//   const prefix = this.props.prefix
//   const field: any = get(this, "props.field")
//   const form = getListFormFromRelationshipField(field)
//   const preppedPrefix = prefix && prefix.length > 0 ? prefix + "." : ""
//   const value = get(this, "props.value", "")
//   if (value.length > 0 && !isNil(form.sections[0].fields[1])) {
//     FormStore.setValue(preppedPrefix + field.mapTo + "." + form.sections[0].fields[1].mapTo,
//       {$nin: Array.isArray(value) ? value : [value]})
//   }
// }

const useRelationship = (schema: ExtendedJSONSchema, mapTo?: string, prefix?: string) => {
  if (mapTo == null) { mapTo = getMapTo(schema, mapTo) }
  const hookedField = useFieldWithValidation(schema, mapTo, prefix)
  const value: any = hookedField.value
  const collection: any = get(schema.relationship, "collection", "")
  // TODO: use prefix
  const displayMode = FormStore.getValue(`displayMode.${mapTo}`)
  let relatedObject: any
  const searchForm = getListFormFromRelationshipField(schema)
  if (isArray(value)) {
    relatedObject = []
    for (const id of value) {
      if (!isNil(id)) {
        relatedObject.push(DbStore.get(id, collection, false))
      }
    }
  } else if (isNumber(value) || isString(value) && value.length > 0) {
    const arg: any = value
    relatedObject = DbStore.get(arg, collection, false)
  }
  const lastUsed = FormStore.getValue(`${lastUsedKey}.${prependPrefix(mapTo, prefix)}`)
  // becausae autoSelect is the only relevant property in schema we only check against this instead of the stringifed schema for speed
  const autoSelect = get(schema.relationship, "autoSelect")
  return {
    value, setValue: hookedField.setValue, relatedObject, searchForm, displayMode, lastUsed,
    addId: memo(() => addId(schema, mapTo, prefix), ["addId", mapTo, prefix, autoSelect]), prefix,
    removeId: memo(() => removeId(schema, mapTo, prefix), ["removeId", mapTo, prefix, autoSelect]),
    searchRelated: memo(() => searchRelated(schema, mapTo, prefix), ["searchRelated", mapTo, prefix, autoSelect]),
    showDisplay: memo(() => displayModeChanger(String(mapTo), prefix)(1), ["showDisplay", mapTo, prefix]),
    hideDisplay: memo(() => displayModeChanger(String(mapTo), prefix)(0), ["hideDisplay", mapTo, prefix])
  }
}

export interface IRelationshipHookProps {
  schema: ExtendedJSONSchema
  mapTo?: string
  prefix?: string
}

const useRelationshipWithProps = (props: IRelationshipHookProps) => useRelationship(props.schema, props.mapTo, props.prefix)

export default useRelationship
export {
  addId, removeId, searchRelated, getListFormFromRelationshipField, useRelationshipWithProps, displayModeChanger
}
