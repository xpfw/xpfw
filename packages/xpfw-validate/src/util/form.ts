import { isObject } from "lodash"
import { IField, IForm } from "../typeDef"

const getFieldsFromForm: (form: IForm) => IField[] = (form: IForm) => {
  const fields: IField[] = []
  if (isObject(form)) {
    for (const section of form.sections) {
      for (const field of section.fields) {
        fields.push(field)
      }
    }
  }
  return fields
}

export { getFieldsFromForm }
