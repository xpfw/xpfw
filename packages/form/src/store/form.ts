import { get, set } from "lodash"
import { action, observable } from "mobx"
import { prependPrefix } from "../util/prefixMaker"

/**
 * Holds data relevant for Forms such as values and validation errors
 */
export class FormStoreClass {
  /**
   * MobX Observable containing all form data
   */
  @observable
  public formData: any = {}
  /**
   * MobX Observable containing all form errors
   */
  @observable
  public errors: any = {}
  /**
   * MobX Observable containing Loading states
   */
  @observable
  public loading: any = {}

  /**
   * Acquire a value from the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   * @param defaultValue value to return if it is not set in the form data
   */
  public getValue(valuePath?: string, prefix?: string, defaultValue?: any): any {
    return get(this.formData, prependPrefix(valuePath, prefix), defaultValue)
  }

  /**
   * Set a value in the form data
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setValue(valuePath?: string, value?: any, prefix?: string) {
    set(this.formData, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getError(valuePath?: string, prefix?: string): any {
    return get(this.errors, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setError(valuePath?: string, value?: any, prefix?: string) {
    set(this.errors, prependPrefix(valuePath, prefix), value)
  }

  /**
   * Acquire a value from the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  public getLoading(valuePath?: string, prefix?: string): any {
    return get(this.loading, prependPrefix(valuePath, prefix))
  }

  /**
   * Set a value in the form errors
   * @param valuePath key where the value of the field will be saved. uses lodash.get/set string syntax style so subfields, array indices etc. can be used within the string
   * @param value the new value
   * @param prefix prepended to mapTo to allow same mapTo keys to have different values
   */
  @action
  public setLoading(valuePath: string, value: boolean, prefix?: string) {
    set(this.loading, prependPrefix(valuePath, prefix), value)
  }
}

const FormStore = new FormStoreClass()
export default FormStore
