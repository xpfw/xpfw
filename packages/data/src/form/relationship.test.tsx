import { relationshipTest, toJS } from "@xpfw/data-tests"
import { ComponentRegistry } from "@xpfw/form"
import { RelationshipMultiField, RelationshipSingleField } from "@xpfw/form-tests"
import "isomorphic-fetch"
import * as React from "react"
import { IRelationshipHookProps, useRelationshipWithProps } from "./relationship"

const MockedRelationship: React.FunctionComponent<IRelationshipHookProps> = (props) => {
  const relation = useRelationshipWithProps(props)
  return (
    <div>{JSON.stringify(toJS(relation), undefined, 2)}</div>
  )
}

ComponentRegistry.registerComponent(String(RelationshipSingleField.type), MockedRelationship,
RelationshipSingleField.theme)
ComponentRegistry.registerComponent(String(RelationshipMultiField.type), MockedRelationship,
RelationshipMultiField.theme)

relationshipTest()
