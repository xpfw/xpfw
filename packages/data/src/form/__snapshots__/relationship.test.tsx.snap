// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DbStore Create Test: after autoselectable search with option enabled and set automatically 1`] = `
<div>
  {
  "value": "b29999999999999999298765",
  "relatedObject": {
    "guestName": "myText 2",
    "myVal": 422,
    "_id": "b29999999999999999298765"
  },
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "lastUsed": [
    "b29999999999999999298765"
  ],
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: after autoselectable search with option enabled but too many results 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: after autoselectable search without option enabled 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: after reset 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: after reset 2`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: customPresearch for Text 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "prefix": "customPre"
}
</div>
`;

exports[`DbStore Create Test: customPresearch for my 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "prefix": "customPre"
}
</div>
`;

exports[`DbStore Create Test: customPresearch for null 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "prefix": "customPre"
}
</div>
`;

exports[`DbStore Create Test: multi added and loaded 1`] = `
<div>
  {
  "value": [
    "b29999999999999999498765",
    "b29999999999999999798765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 7",
      "myVal": 427,
      "_id": "b29999999999999999798765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "lastUsed": [
    "b29999999999999999798765"
  ]
}
</div>
`;

exports[`DbStore Create Test: multi added but not yet loaded 1`] = `
<div>
  {
  "value": [
    "b29999999999999999498765",
    "b29999999999999999798765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 7",
      "myVal": 427,
      "_id": "b29999999999999999798765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "lastUsed": [
    "b29999999999999999798765"
  ]
}
</div>
`;

exports[`DbStore Create Test: multi after autoselectable search with option enabled but too many results 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: multi after autoselectable search without option enabled 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: multi load with none missing 1`] = `
<div>
  {
  "value": [
    "b29999999999999999398765",
    "b29999999999999999498765",
    "b29999999999999999598765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 3",
      "myVal": 423,
      "_id": "b29999999999999999398765"
    },
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 5",
      "myVal": 425,
      "_id": "b29999999999999999598765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: multi load with one missing 1`] = `
<div>
  {
  "value": [
    "b29999999999999999398765",
    "b29999999999999999498765",
    "b29999999999999999598765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 3",
      "myVal": 423,
      "_id": "b29999999999999999398765"
    },
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 5",
      "myVal": 425,
      "_id": "b29999999999999999598765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: multi removed one 1`] = `
<div>
  {
  "value": [
    "b29999999999999999398765",
    "b29999999999999999498765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 3",
      "myVal": 423,
      "_id": "b29999999999999999398765"
    },
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: multi removed two 1`] = `
<div>
  {
  "value": [
    "b29999999999999999498765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: multiafter autoselectable search with option enabled and set automatically 1`] = `
<div>
  {
  "value": [
    "b29999999999999999398765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 3",
      "myVal": 423,
      "_id": "b29999999999999999398765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "lastUsed": [
    "b29999999999999999398765"
  ],
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: multiafter autoselectable search with option enabled and set automatically twice 1`] = `
<div>
  {
  "value": [
    "b29999999999999999398765",
    "b29999999999999999598765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 3",
      "myVal": 423,
      "_id": "b29999999999999999398765"
    },
    {
      "guestName": "myText 5",
      "myVal": 425,
      "_id": "b29999999999999999598765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "lastUsed": [
    "b29999999999999999598765",
    "b29999999999999999398765"
  ],
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: nothing set yet 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: nothing yet 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: nothing yet 2`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: nothing yet 3`] = `
<div>
  {
  "value": [
    "b29999999999999999498765",
    "b29999999999999999798765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 7",
      "myVal": 427,
      "_id": "b29999999999999999798765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "lastUsed": [
    "b29999999999999999798765"
  ],
  "prefix": ""
}
</div>
`;

exports[`DbStore Create Test: nothing yet 4`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "prefix": "customPre"
}
</div>
`;

exports[`DbStore Create Test: search for Text 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: search for Text 2`] = `
<div>
  {
  "value": [
    "b29999999999999999498765",
    "b29999999999999999798765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 7",
      "myVal": 427,
      "_id": "b29999999999999999798765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "lastUsed": [
    "b29999999999999999798765"
  ],
  "prefix": ""
}
</div>
`;

exports[`DbStore Create Test: search for Text but displaymode for items 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: search for my 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: search for my 2`] = `
<div>
  {
  "value": [
    "b29999999999999999498765",
    "b29999999999999999798765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 7",
      "myVal": 427,
      "_id": "b29999999999999999798765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "lastUsed": [
    "b29999999999999999798765"
  ],
  "prefix": ""
}
</div>
`;

exports[`DbStore Create Test: search for null 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 0,
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: search for null 2`] = `
<div>
  {
  "value": [
    "b29999999999999999498765",
    "b29999999999999999798765"
  ],
  "relatedObject": [
    {
      "guestName": "myText 4",
      "myVal": 424,
      "_id": "b29999999999999999498765"
    },
    {
      "guestName": "myText 7",
      "myVal": 427,
      "_id": "b29999999999999999798765"
    }
  ],
  "searchForm": {
    "title": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "displayMode": 1,
  "lastUsed": [
    "b29999999999999999798765"
  ],
  "prefix": ""
}
</div>
`;

exports[`DbStore Create Test: set to first creation loaded 1`] = `
<div>
  {
  "value": "b29999999999999999198765",
  "relatedObject": {
    "guestName": "myText 1",
    "myVal": 421,
    "_id": "b29999999999999999198765"
  },
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: set to first creation not loaded yet 1`] = `
<div>
  {
  "value": "b29999999999999999098765",
  "relatedObject": {
    "guestName": "myText 0",
    "myVal": 420,
    "_id": "b29999999999999999098765"
  },
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: set to invalid value 1`] = `
<div>
  {
  "value": "INVALID",
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  }
}
</div>
`;

exports[`DbStore Create Test: single search for Text 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: single search for my 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "prefix": "ownedBy"
}
</div>
`;

exports[`DbStore Create Test: single search for null 1`] = `
<div>
  {
  "searchForm": {
    "title": "searchFor.ownedBy",
    "collection": "simpleTestCol",
    "properties": {
      "guestName": {
        "title": "guestName",
        "type": "string"
      },
      "_id": {
        "title": "_id",
        "type": "string"
      }
    },
    "relationship": {}
  },
  "prefix": "ownedBy"
}
</div>
`;
